src/tileset_anims.c

```diff
@@ -0,0 +1,248 @@
+#include "global.h"
+#include "graphics.h"
+#include "palette.h"
+#include "util.h"
+#include "battle_transition.h"
+#include "task.h"
+#include "fieldmap.h"
+
+static EWRAM_DATA struct {
+    const u16 *src;
+    u16 *dest;
+    u16 size;
+} sTilesetDMA3TransferBuffer[20] = {0};
+
+static u8 sTilesetDMA3TransferBufferSize;
+static u16 sPrimaryTilesetAnimCounter;
+static u16 sPrimaryTilesetAnimCounterMax;
+static u16 sSecondaryTilesetAnimCounter;
+static u16 sSecondaryTilesetAnimCounterMax;
+static void (*sPrimaryTilesetAnimCallback)(u16);
+static void (*sSecondaryTilesetAnimCallback)(u16);
+
+static void _InitPrimaryTilesetAnimation(void);
+static void _InitSecondaryTilesetAnimation(void);
+
+extern const u16 *const gUnknown_83A7660[];
+extern const u16 *const gUnknown_83AA654[];
+extern const u16 *const gUnknown_83AB874[];
+extern const u16 *const gUnknown_83ABDB4[];
+extern const u16 *const gUnknown_83AC1E8[];
+extern const u16 *const gUnknown_83AC5F8[];
+extern const u16 *const gUnknown_83AC7C8[];
+extern const u16 *const gUnknown_83AC950[];
+
+static void ResetTilesetAnimBuffer(void)
+{
+    sTilesetDMA3TransferBufferSize = 0;
+    CpuFill32(0, sTilesetDMA3TransferBuffer, sizeof sTilesetDMA3TransferBuffer);
+}
+
+static void AppendTilesetAnimToBuffer(const u16 *src, u16 *dest, u16 size)
+{
+    if (sTilesetDMA3TransferBufferSize < 20)
+    {
+        sTilesetDMA3TransferBuffer[sTilesetDMA3TransferBufferSize].src = src;
+        sTilesetDMA3TransferBuffer[sTilesetDMA3TransferBufferSize].dest = dest;
+        sTilesetDMA3TransferBuffer[sTilesetDMA3TransferBufferSize].size = size;
+        sTilesetDMA3TransferBufferSize ++;
+    }
+}
+
+void TransferTilesetAnimsBuffer(void)
```

Comment:
You may want to create `include/tileset_anims.h` and declare all the non-static functions there just like in pokeemerald for future use.